src/probnum/diffeq/odefiltsmooth/initialize.py

```diff
@@ -0,0 +1,318 @@
+        y0=z0,
+        atol=1e-12,
+        rtol=1e-12,
+        t_eval=t_eval,
+        method=method,
+    )
+
+    ts = sol.t[:num_steps]
+    ys = sol.y[:, :num_steps].T
+
+    initmean = np.zeros(prior.dimension)
+    initmean[0 :: (order + 1)] = z0
+    initmean[1 :: (order + 1)] = f(t0, z0)
+
+    initcov_diag = np.ones(prior.dimension)
+    initcov_diag[0 :: (order + 1)] = SMALL_VALUE
+    initcov_diag[1 :: (order + 1)] = SMALL_VALUE
+
+    if df is not None:
+        if order > 1:
+            initmean[2 :: (order + 1)] = df(t0, z0) @ f(t0, z0)
+            initcov_diag[2 :: (order + 1)] = SMALL_VALUE
+
+    initcov = np.diag(initcov_diag)
+    initcov_cholesky = np.diag(np.sqrt(initcov_diag))
+    initrv = pnrv.Normal(initmean, initcov, cov_cholesky=initcov_cholesky)
+    kalman = pnfs.Kalman(prior, measmod, initrv)
+
+    out = kalman.filtsmooth(ys, ts)
+
+    estimated_initrv = out.state_rvs[0]
+    mean = estimated_initrv.mean
+    std = estimated_initrv.std
+    return mean, std
+
+
+def compute_all_derivatives_via_taylormode(f, z0, t0, order):
+    """Compute derivatives of the initial conditions of an IVP with Taylor-mode
+    automatic differentiation.
+
+    This requires JAX. For an explanation of what happens ``under the hood``, see [1]_.
+
+    References
+    ----------
+    .. [1] Krämer, N. and Hennig, P., Stable implementation of probabilistic ODE solvers,
+       *arXiv:2012.10106*, 2020.
+
+
+    The implementation is inspired by the implementation in
+    https://github.com/jacobjinkelly/easy-neural-ode/blob/master/latent_ode.py
+
+    Parameters
+    ----------
+    f
+        ODE vector field.
+    z0
+        Initial value.
+    t0
+        Initial time point.
+    order
+        Number of derivatives to compute. Corresponds to the order of the prior (for instance the number of times an integrated Brownian motion process is integrated).
+
+    Returns
+    -------
+    array
+        Estimate of the full stack of derivatives
+    array
+        Array full of zeros, as a dummy error estimate of the estimation.
+    """
+
+    try:
+        import jax.numpy as jnp
+        from jax.config import config
+        from jax.experimental.jet import jet
+
+        config.update("jax_enable_x64", True)
+    except ImportError as err:
+        raise ImportError(
+            "Cannot perform Taylor-mode initialisation without optional "
+            "dependencies jax and jaxlib. Try installing them via `pip install jax jaxlib`."
+        ) from err
+
+    def total_derivative(z_t):
+        """Total derivative."""
+        z, t = jnp.reshape(z_t[:-1], z_shape), z_t[-1]
+        dz = jnp.ravel(f(t, z))
+        dt = jnp.array([1.0])
+        dz_t = jnp.concatenate((dz, dt))
+        return dz_t
+
+    z_shape = z0.shape
+    z_t = jnp.concatenate((jnp.ravel(z0), jnp.array([t0])))
+
+    derivs = []
+
+    derivs.extend(z0)
+    if order == 0:
+        all_derivs = pnss.convert_derivwise_to_coordwise(
```

Comment:
the taylormode implementation uses a different state-derivative ordering than we do in probnum. Therefore the conversion function. I found that it was useful to have at a few places (like, here as well as in some tests), so it is public and tested etc.

Reply:
In principle, I'd be happy about using the other representation (in which case we wouldnt need the old representation anymore). It is a bit of a project to change it though. 

I actually had this potential future switch in mind, when I decided to make these conversion functions public

Reply:
Decision: make these "private" methods of `Integrator`; use wherever necessary, but don't make them too public